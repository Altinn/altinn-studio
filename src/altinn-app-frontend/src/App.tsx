import { createTheme, MuiThemeProvider } from '@material-ui/core';
import * as React from 'react';
import { Route, Switch } from 'react-router-dom';
import { AltinnAppTheme } from 'altinn-shared/theme';
import ProcessWrapper from './shared/containers/ProcessWrapper';
import UnknownError from './features/instantiate/containers/UnknownError';
import PartySelection from './features/instantiate/containers/PartySelection';
import {
  startInitialAppTaskQueue,
  startInitialUserTaskQueue,
} from './shared/resources/queue/queueSlice';
import { get } from './utils/networking';
import {
  getEnvironmentLoginUrl,
  refreshJwtTokenUrl,
} from './utils/appUrlHelper';
import { makeGetHasErrorsSelector } from './selectors/getErrors';
import Entrypoint from './features/entrypoint/Entrypoint';
import { useAppDispatch, useAppSelector } from './common/hooks';
import { makeGetAllowAnonymousSelector } from './selectors/getAllowAnonymous';
import { AppWrapper } from '@altinn/altinn-design-system';

const theme = createTheme(AltinnAppTheme);

// 1 minute = 60.000ms
const TEN_MINUTE_IN_MILLISECONDS: number = 60000 * 10;

export const App = () => {
  const dispatch = useAppDispatch();
  const hasErrorSelector = makeGetHasErrorsSelector();
  const hasApiErrors: boolean = useAppSelector(hasErrorSelector);
  const appOidcProvider = useAppSelector(
    (state) => state.applicationSettings?.applicationSettings?.appOidcProvider,
  );
  const lastRefreshTokenTimestamp = React.useRef(0);

  const allowAnonymousSelector = makeGetAllowAnonymousSelector();
  const allowAnonymous = useAppSelector(allowAnonymousSelector);

  const [ready, setReady] = React.useState(false);

  React.useEffect(() => {
    function setUpEventListeners() {
      window.addEventListener('mousemove', refreshJwtToken);
      window.addEventListener('scroll', refreshJwtToken);
      window.addEventListener('onfocus', refreshJwtToken);
      window.addEventListener('keydown', refreshJwtToken);
    }

    function removeEventListeners() {
      window.removeEventListener('mousemove', refreshJwtToken);
      window.removeEventListener('scroll', refreshJwtToken);
      window.removeEventListener('onfocus', refreshJwtToken);
      window.removeEventListener('keydown', refreshJwtToken);
    }

    function refreshJwtToken() {
      const timeNow = Date.now();
      if (
        timeNow - lastRefreshTokenTimestamp.current >
        TEN_MINUTE_IN_MILLISECONDS
      ) {
        lastRefreshTokenTimestamp.current = timeNow;
        get(refreshJwtTokenUrl).catch((err) => {
          // Most likely the user has an expired token, so we redirect to the login-page
          try {
            window.location.href = getEnvironmentLoginUrl(appOidcProvider);
          } catch (error) {
            console.error(err, error);
          }
        });
      }
    }

    if (allowAnonymous !== undefined) {
      // Page is ready to be rendered once allowAnonymous value has been determined
      setReady(true);
    }

    if (allowAnonymous === false) {
      refreshJwtToken();
      dispatch(startInitialUserTaskQueue());
      setUpEventListeners();

      return () => {
        removeEventListeners();
      };
    }
  }, [allowAnonymous, dispatch, appOidcProvider]);

  React.useEffect(() => {
    dispatch(startInitialAppTaskQueue());
  }, [dispatch]);

  if (hasApiErrors) {
    return <UnknownError />;
  }

  if (!ready) {
    return null;
  }

  return (
    <AppWrapper>
      <MuiThemeProvider theme={theme}>
        <Switch>
          <Route
            path='/'
            exact={true}
          >
            <Entrypoint allowAnonymous={allowAnonymous} />
          </Route>
          <Route
            path='/partyselection/:errorCode?'
            exact={true}
            component={PartySelection}
          />
          <Route
            path='/instance/:partyId/:instanceGuid'
            exact={true}
            component={ProcessWrapper}
          />
        </Switch>
      </MuiThemeProvider>
    </AppWrapper>
  );
};
