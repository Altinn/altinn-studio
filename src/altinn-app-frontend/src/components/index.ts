import { createContext } from 'react';

import { AddressComponent as Address } from 'src/components/advanced/AddressComponent';
import { AttachmentListComponent } from 'src/components/base/AttachmentListComponent';
import { ButtonComponent } from 'src/components/base/ButtonComponent';
import { CheckboxContainerComponent } from 'src/components/base/CheckboxesContainerComponent';
import DatepickerComponent from 'src/components/base/DatepickerComponent';
import DropdownComponent from 'src/components/base/DropdownComponent';
import { FileUploadComponent } from 'src/components/base/FileUpload/FileUploadComponent';
import { FileUploadWithTagComponent } from 'src/components/base/FileUpload/FileUploadWithTag/FileUploadWithTagComponent';
import { HeaderComponent } from 'src/components/base/HeaderComponent';
import { ImageComponent } from 'src/components/base/ImageComponent';
import { InputComponent } from 'src/components/base/InputComponent';
import { InstantiationButtonComponent } from 'src/components/base/InstantiationButtonComponent';
import { LikertComponent } from 'src/components/base/LikertComponent';
import { MapComponent } from 'src/components/base/MapComponent';
import { NavigationBar as NavigationBarComponent } from 'src/components/base/NavigationBar';
import { PanelComponent } from 'src/components/base/PanelComponent';
import { ParagraphComponent } from 'src/components/base/ParagraphComponent';
import { PrintButtonComponent } from 'src/components/base/PrintButtonComponent';
import { RadioButtonContainerComponent } from 'src/components/base/RadioButtons/RadioButtonsContainerComponent';
import { TextAreaComponent } from 'src/components/base/TextAreaComponent';
import CustomComponent from 'src/components/custom/CustomWebComponent';
import { NavigationButtons as NavigationButtonsComponent } from 'src/components/presentation/NavigationButtons';
import type { IGenericComponentProps } from 'src/components/GenericComponent';
import type {
  ComponentExceptGroupAndSummary,
  IGrid,
} from 'src/features/form/layout';
import type { IComponentFormData } from 'src/utils/formComponentUtils';

import type { ILanguage } from 'altinn-shared/types';

const components: {
  [Type in ComponentExceptGroupAndSummary]: (props: any) => JSX.Element;
} = {
  Header: HeaderComponent,
  Paragraph: ParagraphComponent,
  Image: ImageComponent,
  Input: InputComponent,
  DatePicker: DatepickerComponent,
  Dropdown: DropdownComponent,
  Checkboxes: CheckboxContainerComponent,
  RadioButtons: RadioButtonContainerComponent,
  TextArea: TextAreaComponent,
  FileUpload: FileUploadComponent,
  FileUploadWithTag: FileUploadWithTagComponent,
  Button: ButtonComponent,
  NavigationButtons: NavigationButtonsComponent,
  InstantiationButton: InstantiationButtonComponent,
  AttachmentList: AttachmentListComponent,
  NavigationBar: NavigationBarComponent,
  Likert: LikertComponent,
  Panel: PanelComponent,
  PrintButton: PrintButtonComponent,
  AddressComponent: Address,
  Custom: CustomComponent,
  Map: MapComponent,
};

export interface IComponentProps extends IGenericComponentProps {
  handleDataChange: (
    value: string,
    key?: string,
    skipValidation?: boolean,
    checkIfRequired?: boolean,
  ) => void;
  getTextResource: (key: string) => React.ReactNode;
  getTextResourceAsString: (key: string) => string;
  language: ILanguage;
  shouldFocus: boolean;
  text: React.ReactNode;
  label: () => JSX.Element;
  legend: () => JSX.Element;
  formData?: IComponentFormData;
  isValid?: boolean;
}

export interface IFormComponentContext {
  grid?: IGrid;
  baseComponentId?: string;
}

export const FormComponentContext = createContext<IFormComponentContext>({
  grid: undefined,
  baseComponentId: undefined,
});

export default components;
