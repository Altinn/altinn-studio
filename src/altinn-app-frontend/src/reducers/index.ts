import { combineReducers } from 'redux';
import optionsSlice from '../shared/resources/options/optionsSlice';
import formDataSlice from '../features/form/data/formDataSlice';
import formRulesSlice from 'src/features/form/rules/rulesSlice';
import instantiationSlice from 'src/features/instantiate/instantiation/instantiationSlice';
import applicationMetadataSlice from 'src/shared/resources/applicationMetadata/applicationMetadataSlice';
import attachmentSlice from 'src/shared/resources/attachments/attachmentSlice';
import instanceDataSlice from 'src/shared/resources/instanceData/instanceDataSlice';
import orgsSlice from 'src/shared/resources/orgs/orgsSlice';
import partySlice from '../shared/resources/party/partySlice';
import processSlice from 'src/shared/resources/process/processSlice';
import profileSlice from 'src/shared/resources/profile/profileSlice';
import textResourcesSlice from '../shared/resources/textResources/textResourcesSlice';
import { appApi } from 'src/services/AppApi';
import formDynamicsSlice from '../features/form/dynamics/formDynamicsSlice';
import formLayoutSlice from '../features/form/layout/formLayoutSlice';
import formDataModelSlice from '../features/form/datamodel/datamodelSlice';
import validationSlice from '../features/form/validation/validationSlice';
import isLoadingSlice from '../shared/resources/isLoading/isLoadingSlice';
import languageSlice from '../shared/resources/language/languageSlice';
import queueSlice from '../shared/resources/queue/queueSlice';
import applicationSettingsSlice from '../shared/resources/applicationSettings/applicationSettingsSlice';

const reducers = {
  [applicationMetadataSlice.name]: applicationMetadataSlice.reducer,
  [attachmentSlice.name]: attachmentSlice.reducer,
  [formDataSlice.name]: formDataSlice.reducer,
  [formDataModelSlice.name]: formDataModelSlice.reducer,
  [formDynamicsSlice.name]: formDynamicsSlice.reducer,
  [formLayoutSlice.name]: formLayoutSlice.reducer,
  [formRulesSlice.name]: formRulesSlice.reducer,
  [validationSlice.name]: validationSlice.reducer,
  [instanceDataSlice.name]: instanceDataSlice.reducer,
  [instantiationSlice.name]: instantiationSlice.reducer,
  [isLoadingSlice.name]: isLoadingSlice.reducer,
  [languageSlice.name]: languageSlice.reducer,
  [orgsSlice.name]: orgsSlice.reducer,
  [partySlice.name]: partySlice.reducer,
  [processSlice.name]: processSlice.reducer,
  [profileSlice.name]: profileSlice.reducer,
  [queueSlice.name]: queueSlice.reducer,
  [textResourcesSlice.name]: textResourcesSlice.reducer,
  [optionsSlice.name]: optionsSlice.reducer,
  [applicationSettingsSlice.name]: applicationSettingsSlice.reducer,
  [appApi.reducerPath]: appApi.reducer,
};

export default combineReducers(reducers);
