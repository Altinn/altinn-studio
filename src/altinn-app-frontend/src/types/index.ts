import { IIsLoadingState } from 'src/shared/resources/isLoading/isLoadingReducers';
import { IOptionsState } from 'src/shared/resources/options/optionsReducer';
import { IFormRuleState } from 'src/features/form/rules/rulesReducer';
import ajv from 'ajv';
import { IFormDataState } from '../features/form/data/formDataReducer';
import { IDataModelState } from '../features/form/datamodel/formDatamodelReducer';
import { IFormDynamicState } from '../features/form/dynamics';
import { ILayoutState } from '../features/form/layout/formLayoutReducer';
import { IValidationState } from '../features/form/validation/validationReducer';
import { IInstantiationState } from '../features/instantiate/instantiation/reducer';
import { IApplicationMetadataState } from '../shared/resources/applicationMetadata/reducer';
import { IAttachmentState } from '../shared/resources/attachments/attachmentReducer';
import { IInstanceDataState } from '../shared/resources/instanceData/instanceDataReducers';
import { ILanguageState } from '../shared/resources/language/languageReducers';
import { IOrgsState } from '../shared/resources/orgs/orgsReducers';
import { IPartyState } from '../shared/resources/party/partyReducers';
import { IProcessState } from '../shared/resources/process/processReducer';
import { IProfileState } from '../shared/resources/profile/profileReducers';
import { IQueueState } from '../shared/resources/queue/queueReducer';
import { ITextResourcesState } from '../shared/resources/textResources/textResourcesReducer';

export type FormComponentType =
  | IFormAddressComponent
  | IFormButtonComponent
  | IFormCheckboxComponent
  | IFormComponent
  | IFormDropdownComponent
  | IFormFileUploaderComponent
  | IFormHeaderComponent
  | IFormInputComponent
  | IFormRadioButtonComponent
  | IFormTextAreaComponent;

export interface IAltinnWindow extends Window {
  app: string;
  conditionalRuleHandlerHelper: IRules;
  instanceId: string;
  org: string;
  reportee: string;
}

export interface IComponentBindingValidation {
  errors?: string[];
  warnings?: string[];
}

export interface IComponentValidations {
  [id: string]: IComponentBindingValidation;
}

export interface IDataModelBinding {
  fieldName: string;
  parentGroup: string;
}

export interface IDataModelBindings {
  [id: string]: string;
}

export interface IDataModelFieldElement {
  choices?: any;
  customProperties?: any;
  dataBindingName: string;
  displayString: string;
  fixedValue?: any;
  id: string;
  isReadOnly: boolean;
  isTagContent: boolean;
  jsonSchemaPointer: string;
  maxOccurs: number;
  minOccurs: number;
  name: string;
  parentElement: string;
  restrictions: any;
  texts: any;
  type: string;
  typeName?: string;
  xmlSchemaXPath: string;
  xName?: string;
  xPath: string;
  xsdValueType?: string;
}

export interface IFormAddressComponent extends IFormComponent {
  simplified: boolean;
}

export interface IFormButtonComponent extends IFormComponent {
  onClickAction: () => void;
}

export interface IFormCheckboxComponent extends IFormComponent {
  options: IOption[];
  preselectedOptionIndex?: number;
}

export interface IFormComponent {
  id: string;
  disabled?: boolean;
  required?: boolean;
  readOnly?: boolean;
}

export interface IFormDropdownComponent extends IFormComponent {
  options: IOption[];
  optionsId: string;
}

export interface IFormFileUploaderComponent extends IFormComponent {
  description: string;
  hasCustomFileEndings: boolean;
  maxFileSizeInMB: number;
  displayMode: string;
  maxNumberOfAttachments: number;
  minNumberOfAttachments: number;
  validFileEndings?: string;
}

export interface IFormHeaderComponent extends IFormComponent {
  size: string;
}

export interface IFormInputComponent extends IFormComponent {
  type: string;
  disabled?: boolean;
}

export interface IFormRadioButtonComponent extends IFormComponent {
  options: IOption[];
  preselectedOptionIndex?: number;
}

export interface IFormTextAreaComponent extends IFormComponent { }

export interface ILayoutSettings {
  pages: IPagesSettings;
}

export interface IPagesSettings {
  order: string[];
}

export interface ILayoutNavigation {
  next?: string;
  previous?: string;
}

export interface INavigationConfig {
  [id: string]: ILayoutNavigation;
}

export interface IOption {
  label: string;
  value: any;
}

export interface IOptions {
  [id: string]: IOption[];
}

export interface IRepeatingGroup {
  count: number;
}

export interface IRepeatingGroups {
  [id: string]: IRepeatingGroup;
}

export interface IRules {
  [id: string]: any;
}

export interface IRuntimeStore {
  attachments: IAttachmentState;
  formData: IFormDataState;
  formDataModel: IDataModelState;
  formDynamics: IFormDynamicState;
  formLayout: ILayoutState;
  language: ILanguageState;
}

export interface IRuntimeState {
  applicationMetadata: IApplicationMetadataState;
  attachments: IAttachmentState;
  formData: IFormDataState;
  formDataModel: IDataModelState;
  formDynamics: IFormDynamicState;
  formLayout: ILayoutState;
  formRules: IFormRuleState;
  formValidations: IValidationState;
  instanceData: IInstanceDataState;
  instantiation: IInstantiationState;
  isLoading: IIsLoadingState;
  language: ILanguageState;
  optionState: IOptionsState;
  organisationMetaData: IOrgsState;
  party: IPartyState;
  process: IProcessState;
  profile: IProfileState;
  queue: IQueueState;
  textResources: ITextResourcesState;
}

export interface ISchemaValidator {
  rootElement: any;
  rootElementPath: string;
  schema: any;
  validator: ajv.Ajv;
}

export interface ITextResource {
  id: string;
  value: string;
  unparsedValue: string;
  variables:IVariable[];
}

export interface ITextResourceBindings {
  [id: string]: string;
}

export interface IValidationIssue {
  code: string;
  description: string;
  field: string;
  scope: string;
  severity: Severity;
  targetId: string;
}

export interface IUiConfig {
  autoSave: boolean;
  currentView: string;
  focus: string;
  hiddenFields: string[];
  repeatingGroups?: IRepeatingGroups;
  navigationConfig?: INavigationConfig;
  layoutOrder: string[];
}

export interface IValidationResult {
  invalidDataTypes: boolean;
  validations: IValidations;
}

export interface IValidations {
  [id: string]: IComponentValidations;
}

export interface IVariable {
  dataSource: string;
  key: string;
}

export enum ProcessSteps {
  Unknown = 'unknown',
  FormFilling = 'data',
  Archived = 'ended',
  Confirm = 'confirmation',
  Feedback = 'feedback',
}

export enum Severity {
  Unspecified = 0,
  Error = 1,
  Warning = 2,
  Informational = 3,
}

export enum Triggers {
  Validation = 'validation',
}
