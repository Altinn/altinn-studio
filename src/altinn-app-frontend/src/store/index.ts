import createSagaMiddleware, { SagaMiddleware } from 'redux-saga';
import reducers from '../reducers';
import { appApi } from 'src/services/AppApi';
import { configureStore, PreloadedState } from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';

export const sagaMiddleware: SagaMiddleware<any> = createSagaMiddleware();
const middlewares = [sagaMiddleware, appApi.middleware];

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  const innerStore = configureStore({
    reducer: reducers,
    devTools: process.env.NODE_ENV !== 'production',
    middleware: (getDefaultMiddleware) => getDefaultMiddleware({
      // TODO: enable once we have cleaned up our store
      serializableCheck: false,
      immutableCheck: true,
    }).concat(middlewares),
    preloadedState
  });

  setupListeners(innerStore.dispatch);
  return innerStore;
};

export const store = setupStore();

// Expose store when running in Cypress. This allows for using cy.getReduxState() to run assertions against the redux
// state at various points in the tests. Testing the state directly might expose problems not easily/visibly testable
// in the app itself.
if ((window as any).Cypress) {
  (window as any).reduxStore = store;
}

export type RootState = ReturnType<typeof reducers>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
