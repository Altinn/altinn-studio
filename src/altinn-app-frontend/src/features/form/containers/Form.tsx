/* eslint-disable no-undef */
import Grid from '@material-ui/core/Grid';
import * as React from 'react';
import { useSelector } from 'react-redux';
import { SummaryComponent } from 'src/components/summary/SummaryComponent';
import { IRuntimeState } from '../../../types';
import { ILayout, ILayoutComponent, ILayoutGroup } from '../layout';
import { GroupContainer } from './GroupContainer';
import { renderGenericComponent } from '../../../utils/layout';
import { DisplayGroupContainer } from './DisplayGroupContainer';

export function renderLayoutComponent(layoutComponent: ILayoutComponent | ILayoutGroup, layout: ILayout) {
  switch (layoutComponent.type) {
    case 'group':
    case 'Group': {
      return RenderLayoutGroup(layoutComponent as ILayoutGroup, layout);
    }
    case 'Summary': {
      return (
        <SummaryComponent key={layoutComponent.id} {...(layoutComponent as ILayoutComponent)} />
      );
    }
    default: {
      return (
        <RenderGenericComponent key={layoutComponent.id} {...(layoutComponent as ILayoutComponent)} />
      );
    }
  }
}

function RenderGenericComponent(component: ILayoutComponent, layout: ILayout) {
  return renderGenericComponent(component, layout);
}

function RenderLayoutGroup(layoutGroup: ILayoutGroup, layout: ILayout): JSX.Element {
  const groupComponents = layoutGroup.children.map((child) => {
    let childId = child;
    if (layoutGroup.edit?.multiPage) {
      childId = child.split(':')[1] || child;
    }
    return layout.find((c) => c.id === childId) as ILayoutComponent;
  });
  const repeating = layoutGroup.maxCount > 1;
  if (!repeating) {
    // If not repeating, treat as regular components
    return (
      <DisplayGroupContainer
        container={layoutGroup}
        components={groupComponents}
        renderLayoutComponent={renderLayoutComponent}
      />
    );
  }

  return (
    <GroupContainer
      container={layoutGroup}
      id={layoutGroup.id}
      key={layoutGroup.id}
      components={groupComponents}
    />
  );
}

export function Form() {
  const [filteredLayout, setFilteredLayout] = React.useState<any[]>([]);
  const [currentLayout, setCurrentLayout] = React.useState<string>();

  const currentView = useSelector((state: IRuntimeState) => state.formLayout.uiConfig.currentView);
  const layout: ILayout =
    useSelector((state: IRuntimeState) => state.formLayout.layouts[state.formLayout.uiConfig.currentView]);

  React.useEffect(() => {
    setCurrentLayout(currentView);
  }, [currentView]);

  React.useEffect(() => {
    let componentsToRender: any[] = layout;
    let renderedInGroup: string[] = [];
    if (layout) {
      const groupComponents = layout.filter((component) => component.type.toLowerCase() === 'group');
      groupComponents.forEach((component: ILayoutGroup) => {
        let childList = component.children;
        if (component.edit?.multiPage) {
          childList = component.children.map((childId) => childId.split(':')[1] || childId);
        }
        renderedInGroup = renderedInGroup.concat(childList);
      });
    }

    if (layout) {
      componentsToRender = layout.filter((component) => {
        return !renderedInGroup.includes(component.id);
      });
    }
    setFilteredLayout(componentsToRender);
  }, [layout]);

  return (
    <Grid
      container={true}
      spacing={3}
      alignItems='flex-start'
    >
      {currentView === currentLayout && filteredLayout && filteredLayout.map((component) => {
        return renderLayoutComponent(component, layout);
      })}
    </Grid>
  );
}

export default Form;
