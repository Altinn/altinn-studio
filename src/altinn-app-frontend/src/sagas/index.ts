import type { SagaIterator } from 'redux-saga';
import { fork, all } from 'redux-saga/effects';

import { sagaMiddleware } from 'src/store';
import { rootSagas } from 'src/shared/resources/utils/sagaSlice';

function* root(): SagaIterator {
  yield all(rootSagas.map((saga) => fork(saga)));
}

export const initSagas = () => sagaMiddleware.run(root);
