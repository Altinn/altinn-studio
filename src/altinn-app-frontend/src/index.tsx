import * as React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { HashRouter } from 'react-router-dom';
import App from './App';
import { initSagas } from './sagas';
import { store } from './store';
import 'core-js/es';
import './styles/index.css';

import ErrorBoundary from './components/ErrorBoundary';

initSagas();

render(
  <Provider store={store}>
    <HashRouter>
      <ErrorBoundary>
        <App />
      </ErrorBoundary>
    </HashRouter>
  </Provider>,
  document.getElementById('root'),
);
