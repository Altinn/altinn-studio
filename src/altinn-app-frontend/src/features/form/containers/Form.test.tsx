import * as React from 'react';
import { Form } from './Form';
import type { PreloadedState } from '@reduxjs/toolkit';
import type { RootState } from 'src/store';
import { renderWithProviders, mockMediaQuery } from 'src/../testUtils';
import type { ILayout, ILayoutComponent } from '../layout';
import { getFormLayoutStateMock } from 'src/../__mocks__/formLayoutStateMock';
import { screen, within } from '@testing-library/react';

const { setScreenWidth } = mockMediaQuery(992);

describe('Form.tsx', () => {
  beforeAll(() => {
    // Set screen size to desktop
    setScreenWidth(1200);
  });

  const mockComponents: ILayout = [
    {
      id: 'field1',
      type: 'Input',
      dataModelBindings: {
        simpleBinding: 'Group.prop1',
      },
      textResourceBindings: {
        title: 'First title',
      },
      readOnly: false,
      required: true,
      disabled: false,
    },
    {
      id: 'field2',
      type: 'Input',
      dataModelBindings: {
        simpleBinding: 'Group.prop2',
      },
      textResourceBindings: {
        title: 'Second title',
      },
      readOnly: false,
      required: false,
      disabled: false,
    },
    {
      id: 'field3',
      type: 'Input',
      dataModelBindings: {
        simpleBinding: 'Group.prop3',
      },
      textResourceBindings: {
        title: 'Third title',
      },
      readOnly: false,
      required: false,
      disabled: false,
    },
    {
      id: 'testGroupId',
      type: 'group',
      dataModelBindings: {
        group: 'Group',
      },
      maxCount: 3,
      children: ['field1', 'field2', 'field3'],
    },
  ];

  it('should render components and groups', () => {
    renderForm();
    expect(screen.getByText('First title')).toBeInTheDocument();
    expect(screen.getByText('Second title')).toBeInTheDocument();
    expect(screen.getByText('Third title')).toBeInTheDocument();
  });

  it('should render DisplayGroupContainer and children if group is non repeating', () => {
    const layoutWithNonRepGroup: ILayout = [
      ...mockComponents,
      {
        id: 'non-rep-group-id',
        type: 'group',
        dataModelBindings: {
          group: 'Group',
        },
        children: ['non-rep-child'],
      },
      {
        id: 'non-rep-child',
        type: 'Input',
        dataModelBindings: {
          simpleBinding: 'Group.prop3',
        },
        textResourceBindings: {
          title: 'Title from non repeating child',
        },
        readOnly: false,
        required: false,
        disabled: false,
      },
    ];

    renderForm(layoutWithNonRepGroup);
    const container = screen.getByTestId('display-group-container');
    expect(container).toBeInTheDocument();
    expect(
      within(container).getByText('Title from non repeating child'),
    ).toBeInTheDocument();
  });

  it('should render PanelGroupContainer and children if group has panel prop', () => {
    const layoutWithPanelGroup: ILayout = [
      ...mockComponents,
      {
        id: 'panel-group-id',
        type: 'group',
        dataModelBindings: {
          group: 'Group',
        },
        children: ['panel-group-child'],
        panel: {
          variant: 'info',
        },
      },
      {
        id: 'panel-group-child',
        type: 'Input',
        dataModelBindings: {
          simpleBinding: 'Group.prop3',
        },
        textResourceBindings: {
          title: 'Title from panel child',
        },
        readOnly: false,
        required: false,
        disabled: false,
      },
    ];

    renderForm(layoutWithPanelGroup);
    const container = screen.getByTestId('panel-group-container');
    expect(container).toBeInTheDocument();
    expect(
      within(container).getByText('Title from panel child'),
    ).toBeInTheDocument();
  });

  it('should render navbar', () => {
    const layoutWithNavBar: ILayout = [
      ...mockComponents,
      {
        id: 'navBar',
        type: 'NavigationBar',
      } as ILayoutComponent,
    ];
    renderForm(layoutWithNavBar);
    expect(screen.getByRole('navigation')).toBeInTheDocument();
    expect(screen.getByText('1. FormLayout')).toBeInTheDocument();
  });

  function renderForm(
    layout = mockComponents,
    customState: PreloadedState<RootState> = {},
  ) {
    renderWithProviders(<Form />, {
      preloadedState: {
        ...customState,
        formLayout: getFormLayoutStateMock({
          layouts: {
            FormLayout: layout,
          },
        }),
      },
    });
  }
});
