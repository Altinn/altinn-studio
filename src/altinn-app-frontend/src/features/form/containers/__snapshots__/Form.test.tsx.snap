// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> features/form/components/Form.tsx +++ should match snapshot 1`] = `
<div>
  <div
    className="MuiGrid-root makeStyles-banner-1 makeStyles-default-2 MuiGrid-item"
    data-testid="MessageBanner-container"
    id="MessageBanner-container"
  >
    <span>
      Obligatoriske felter er markert med *
    </span>
  </div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-align-items-xs-flex-start"
  />
</div>
`;
